wdl-lint/src/rules/runtime_section_keys.rs

```diff
@@ -0,0 +1,355 @@
+//! A lint rule for the `runtime` section keys.
+//!
+//! Note that this lint rule will only emit diagnostics for WDL documents that
+//! have a major version of 1 but a minor version of less than 2, as the
+//! `runtime` section was deprecated in WDL v1.2.
+
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::ops::Deref;
+
+use lazy_static::lazy_static;
+use wdl_ast::span_of;
+use wdl_ast::v1::RuntimeItem;
+use wdl_ast::v1::RuntimeSection;
+use wdl_ast::v1::TaskDefinition;
+use wdl_ast::version::V1;
+use wdl_ast::AstToken;
+use wdl_ast::Diagnostic;
+use wdl_ast::Diagnostics;
+use wdl_ast::Span;
+use wdl_ast::SupportedVersion;
+use wdl_ast::VisitReason;
+use wdl_ast::Visitor;
+
+use crate::Rule;
+use crate::Tag;
+use crate::TagSet;
+
+/// The identifier for the runtime section rule.
+const ID: &str = "RuntimeSectionKeys";
+
+/// A deprecated key.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct DeprecatedKey {
+    /// The key name.
+    name: &'static str,
+    /// The equivalent key that should be used instead.
+    replacement: &'static str,
+}
+
+/// A runtime key and the corresponding status of that key.
+///
+/// These are intended to be assigned at a per-version level of granularity.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+enum Key {
+    /// A key that is deprecated in favor of another key.
+    Deprecated(DeprecatedKey),
+    /// A runtime key that is recommended to be included.
+    Recommended(&'static str),
+    /// A runtime key that has a reserved meaning in the specification but which
+    /// execution engines are _not_ required to support. These are also called
+    /// "hints" in WDL parlance.
+    Hint(&'static str),
+    /// A runtime key that has a reserved meaning in the specification and which
+    /// execution engines are required to support (but don't necessarily have to
+    /// be present in WDL documents).
+    ReservedMandatory(&'static str),
```

Comment:
It's a good question and I considered that briefly too. To me, since this will be hidden as a behind-the-scenes implementation detail combined with the fact that the specification refers to them as "reserved, mandatory keys", I felt it was okay to go with this.